feat(api): sftp project, part 1

RamilGaripov opened the pull request:
refs. metriport/metriport-internal#2791

### Description
- Added `patient_settings` table
- Patient create event also creates a row in `patient_settings` 
- Patient delete event also deletes a row in `patient_settings` 
- Added internal route to bulk update patient settings for a CX
  - Created a new endpoint on Postman 

### Testing

- Local
  - [x] Create new patient, check `patient_settings`
  - [x] Delete new patient, check `patient_settings`
  - [x] Created/updated settings using different combinations on the Postman endpoint
- Staging
  - [x] Create new patient, check `patient_settings`
  - [x] Delete new patient, check `patient_settings`
  - [x] Created/updated settings using different combinations on the Postman endpoint
  - [x] Create and then update patient settings for >8k patients 🚀 
- Production
  - [ ] Monitor that new patient creates result in new rows on `patient_settings`
  - [ ] Create/update settings thru the Postman endpoint for a random patient

### Release Plan

- :warning: This contains a DB migration
- [ ] Release NPM packages
- [ ] Merge this

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Patient management now supports integrated settings, enabling concurrent updates during creation, update, and deletion.
  - New endpoints allow bulk processing of patient settings.
  - Introduced a new patient settings model and related schemas for better data management.
  - Enhanced handling of patient settings during patient creation and deletion processes.
  - New functionalities for managing patient settings, including creation, upsertion, and deletion.
  - Added optional settings parameter during patient creation for enhanced customization.
  - New router for managing patient settings with dedicated endpoints.
  - Updated patient creation schema to include optional settings.
  - New import and integration for patient settings within existing routes.

- **Bug Fixes**
  - Resolved minor typographical issues to ensure accurate linking of patient identifiers.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

leite08 commented on packages/api/src/command/medical/patient/create-patient-settings.ts:
```diff
@@ -0,0 +1,157 @@
+}> {
+  const { log } = out(`createOrUpdatePatientSettings - cx ${cxId}`);
+
+  const idsToProcess = Array.isArray(patientIds) ? patientIds : [];
```
The param is strongly typed, we should check that at the route level.

leite08 commented on packages/api/src/command/medical/patient/create-patient-settings.ts:
```diff
@@ -0,0 +1,157 @@
+  const foundPatientIds = new Set(patients.map(p => p.id));
+  const notFoundIds = patientIds.filter(id => !foundPatientIds.has(id));
+  return {
+    validPatientIds: Array.from(foundPatientIds),
+    notFoundIds,
```
Let's name these properties more similar one to the other? Like, "validIds" and "invalidIds" or "foundPatientIds" and "notFoundPatientIds", etc.

lucasdellabella replied:
+1

leite08 commented on packages/api/src/command/medical/patient/create-patient-settings.ts:
```diff
@@ -0,0 +1,157 @@
+ */
+export async function createOrUpdatePatientSettings({
+  cxId,
+  patientIds = [],
```
Since we're defaulting this to all patients, maybe we add the option to set the facility so we can filter the patient IDs like that?

RamilGaripov replied:
Idk if we need that. This is an internal route, and we can pass the IDs that we want anyway, so this is unnecessary

lucasdellabella replied:
Yeah agreed with Ramil here, this is YAGNI, would prefer to not add additional functionality to an endpoint unless its actually required.

I think its an interesting proposal, as there may be situations where we only want to enable e.g. Circle Arizona based patients, but thats something that shouldn't be tackled rn imo.

leite08 replied:
We need to strive for a balance between building explicitly what we need right now and what is cheap to build and we think we'll need soon.

It's quite common to have requests based on the facility ID and this is super simple to add, that's why I suggested it.

leite08 replied:
Not blocking and NABD, though.

RamilGaripov replied:
Added it anyway

leite08 replied:
Nice! Just a nit: add it to the TSDoc

leite08 commented on packages/api/src/command/medical/patient/create-patient-settings.ts:
```diff
@@ -0,0 +1,157 @@
+  const { log } = out(`createOrUpdatePatientSettings - cx ${cxId}`);
+
+  const idsToProcess = Array.isArray(patientIds) ? patientIds : [];
+  let finalIds = idsToProcess.length > 0 ? idsToProcess : await getPatientIds({ cxId });
```
Why creating this as a variable? We can easily create a const here and a new array to get the resulting, validated patient IDs, so we can make the code more immutable (deterministic).

leite08 commented on packages/api/src/command/medical/patient/create-patient-settings.ts:
```diff
@@ -0,0 +1,157 @@
+  let patientsUpdated = 0;
+
+  await executeOnDBTx(PatientModel.prototype, async transaction => {
+    // Only verify patients if specific IDs were provided
```
😍

leite08 commented on packages/api/src/command/medical/patient/create-patient-settings.ts:
```diff
@@ -0,0 +1,157 @@
+  let patientsNotFound: string[] = [];
+  let patientsUpdated = 0;
+
+  await executeOnDBTx(PatientModel.prototype, async transaction => {
```
Why do we need to run this inside a DB tx?

leite08 commented on packages/api/src/command/medical/patient/create-patient-settings.ts:
```diff
@@ -0,0 +1,157 @@
+  };
+
+  const newPatientSettings = await PatientSettingsModel.create(patientSettingsCreate);
+  return newPatientSettings;
```
return `.dataValues` so we're not returning the Model, please

lucasdellabella replied:
+1+1+1

leite08 commented on packages/api/src/command/medical/patient/create-patient-settings.ts:
```diff
@@ -0,0 +1,157 @@
+}): Promise<PatientSettings> {
+  await getPatientOrFail({ cxId, id: patientId });
+
+  const patientSettingsCreate: PatientSettingsCreate = {
```
The type here can be `PatientSettings`, right? We're providing the `id`.

leite08 commented on packages/api/src/command/medical/patient/create-patient-settings.ts:
```diff
@@ -0,0 +1,157 @@
+  patientId,
+  cxId,
+  adtSubscription,
+}: {
+  patientId: string;
+  cxId: string;
+  adtSubscription: boolean;
+}): Promise<PatientSettings> {
```
The input type is `PatientSettingsCreate`, right?

RamilGaripov replied:
Done

leite08 commented on packages/core/src/domain/patient-settings.ts:
```diff
@@ -0,0 +1,10 @@
+import { BaseDomain, BaseDomainCreate } from "./base-domain";
+
+export interface PatientSettingsCreate extends BaseDomainCreate {
+  id: string;
```
Could move this to `PatientSettings` so the "create" represents the input to the create command function.

leite08 replied:
I don't see it on the PR, maybe forgot to push the commit?

leite08 replied:
Actually, we don't need this here as the `id` is already part of the `BaseDomainCreate`.

leite08 commented on packages/api/src/models/patient-settings.ts:
```diff
@@ -0,0 +1,37 @@
+        },
+        adtSubscription: {
+          type: DataTypes.BOOLEAN,
+          allowNull: false,
```
I think we only set these on the migration, and then we do it for all columns.

leite08 commented on packages/api/src/sequelize/migrations/2025-03-13_00_create-patient-settings.ts:
```diff
@@ -0,0 +1,45 @@
+          type: DataTypes.BOOLEAN,
+          field: adtSubscriptionColumn,
+          allowNull: false,
+          defaultValue: false,
```
Since we have the default value on the command, we could skip this - otherwise there's no single source of truth to it.

leite08 replied:
I think you mean the opposite, as I see on the last version of the PR, right?

RamilGaripov replied:
I'm a little confused about this. Hope the current state is ok? Lmk if not

leite08 commented on packages/api/src/sequelize/migrations/2025-03-13_00_create-patient-settings.ts:
```diff
@@ -0,0 +1,45 @@
+      patientSettingsTableName,
+      {
+        id: {
+          type: DataTypes.STRING,
```
We use `DataTypes.UUID` for ID columns. Any particular reason to use `STRING`?

leite08 commented on packages/api/src/command/medical/patient/create-patient.ts:
```diff
@@ -26,12 +27,14 @@ export async function createPatient({
   rerunPdOnNewDemographics,
   forceCommonwell,
   forceCarequality,
+  adtSubscription = false,
```
We already have this default to `false` on `createPatientSettings`, no need to duplicate this logic it seems.

leite08 replied:
We stil have at least 3 places setting adt to false by default. Maybe could create a const at the top of the file so we only define it once?

RamilGaripov replied:
Yup, created that

leite08 commented on packages/api/src/command/medical/patient/create-patient-settings.ts:
```diff
@@ -0,0 +1,157 @@
+  validPatientIds: string[];
+  notFoundIds: string[];
+}> {
+  const patients = await PatientModel.findAll({
+    where: { id: patientIds, cxId },
+    attributes: ["id"],
+    transaction,
+  });
```
NABD, but could call `getPatientIds` if we add the array of patient IDs as an optional parameter.

leite08 replied:
I don't understand what you mean :)

I meant replacing this

```ts
  const patients = await PatientModel.findAll({
    where: { id: patientIds, cxId },
    attributes: ["id"],
  });
```

...by this

```ts
  const patients = await getPatientIds({ cxId, patientIds });
```

But defs NABD and we can leave as is, since `getPatientIds` getting a list of patient IDs looks goofy.

RamilGaripov replied:
Ah I see. Ya, I'll leave as is if cool

leite08 commented on packages/api/src/command/medical/patient/create-patient-settings.ts:
```diff
@@ -0,0 +1,157 @@
+}): Promise<{
+  validPatientIds: string[];
+  notFoundIds: string[];
+}> {
+  const patients = await PatientModel.findAll({
```
nit: no need to hit the DB if the array of patient IDs is empty. I know the current flow doesn't allow this situation, but the function could be reused as we update the codebase. Defs NADB, though.

RamilGaripov replied:
Addressed to the best of my understanding

leite08 reviewed (commented):
Missing update:
- SDK
- Docs

RamilGaripov commented on packages/api/src/routes/medical/patient-root.ts:
```diff
@@ -57,6 +57,7 @@ router.post(
     );
     const forceCommonwell = stringToBoolean(getFrom("query").optional("commonwell", req));
     const forceCarequality = stringToBoolean(getFrom("query").optional("carequality", req));
+    const adtSubscription = stringToBoolean(getFrom("query").optional("adtSubscription", req));
```
I think I prefer a more structured way, like:
```
"subscribeTo": {
    "adt": true
 }
 ```

leite08 replied:
I didn't mean an array of strings, I meant as a new property inside the body, along w/ the demographics.

I see how looking from the perspective of a single param it makes sense to have this as a query param, but we need to think about the API as it evolves through time, how we minimize changes, and how we make it easy/straightforward to cxs.

Like, it's much easier to look at the documentation and see a property that relates to all subscriptions I can enable a patient to, than separate params, with no common prefix.

Could even call it "settings", and have the "subscription" sub-properties inside it.

This is prob a good topic to bring to #engineering, it's larger than the scope of this PR - but since we're touching it here, we need to discuss it.

RamilGaripov replied:
Yea, I went with a larger settings attribute that has subscriptions as one (and only) property in it for now

lucasdellabella commented on packages/api/src/command/medical/patient/create-patient-settings.ts:
```diff
@@ -0,0 +1,157 @@
+ * @param adtSubscription Whether to enable or disable ADT subscription. Optional, defaults to false.
+ * @returns The number of patients updated and the list of patients not found
+ */
+export async function createOrUpdatePatientSettings({
```
This is usually known as an upsert - I'd like for us to use that term here unless there's a reason we're avoiding it.

upsertPatientSettings()

lucasdellabella commented on packages/api/src/command/medical/patient/create-patient-settings.ts:
```diff
@@ -0,0 +1,157 @@
+ * @param adtSubscription Whether to enable or disable ADT subscription. Optional, defaults to false.
+ * @returns The number of patients updated and the list of patients not found
+ */
+export async function createOrUpdatePatientSettings({
```
I think we chatted about this, but I'd like to break this route out into two routes,

upsertPatientSettings() and upsertAllPatientSettingsForCx() or similar. The current approach technically works fine, but 
1. imo API design is all about managing developer expectations. It would be _very_ surprising to any user of this API that providing an empty array of patients for a CX would instead update all patients. 
2. We want to avoid branching logic! One of the things that makes code hard to read / follow is having branching logic. Having branching logic inside the endpoint makes shit more complicated than if we were to have the caller decide which endpoint to invoke.


It also makes it more clear how to work with these endpoints if `upsertAllPatientSettingsForCx()` simply has no patientIds parameter, see below:

```
upsertPatientSettings(
  cxId
  patientIds
  settings
)

upsertAllPatientSettingsForCx(
  cxId
  settings
)
```

lucasdellabella replied:
awesome

lucasdellabella reviewed (changes requested):
Main thing is breaking out into 2 api endpoints imo

lucasdellabella commented on packages/api/src/command/medical/patient/create-patient-settings.ts:
```diff
@@ -18,139 +20,127 @@ import { getPatientIds, getPatientOrFail } from "./get-patient";
   await getPatientOrFail({ cxId, id: patientId });
 
   const patientSettingsCreate: PatientSettingsCreate = {
     id: uuidv7(),
```
DB should be autogenerating this id on insert, I don't think this is needed, right? Sorry that I missed this earlier.

RamilGaripov replied:
I'm not sure how to make that work. The BaseDomain expects an "id". The normal PatientCreate has this kinda implementation (which I copied). Cool if we stick with it?

lucasdellabella replied:
If this is convention already, fine by me (cc @leite08)

leite08 replied:
Yeah, we've been setting the ID in memory. This is fine b/c we use UUIDs, and the DB implementation doesn't use v7 AFAIK.

lucasdellabella commented on packages/api/src/sequelize/migrations/2025-03-13_00_create-patient-settings.ts:
```diff
@@ -12,34 +14,46 @@ export const up: Migration = async ({ context: queryInterface }) => {
-          type: DataTypes.STRING,
+          type: DataTypes.UUID,
           primaryKey: true,
           allowNull: false,
```
autogenerate the uuid here, yea?

RamilGaripov replied:
we want to generate uuidv7 in mem, as stated in another thread

lucasdellabella commented on packages/utils/src/consolidated/create-consolidate-local.ts:
Did this accidentally come from a dirty working area? Seems related to other work.

RamilGaripov replied:
Removed

RamilGaripov commented:
> Missing update:
> 
> * SDK
> * Docs

I updated the SDK now, but I think we want to delay updating the docs for now, until the ADT subscription is actually functional. I added the TODO for this in the [Notion doc](https://www.notion.so/metriport/ADT-Notifications-Patient-Roster-SFTP-Lambda-19cb51bb90408025aabafab9863a1420?pvs=4#1b5b51bb9040806e968ded85c910ade5) and the corresponding GH Ticket - [link](https://github.com/metriport/metriport-internal/issues/2791)

RamilGaripov commented on packages/api-sdk/src/medical/client/metriport.ts:
```diff
@@ -340,16 +340,18 @@ export class MetriportMedicalApi {
   async createPatient(
     data: PatientCreate,
     facilityId: string,
+    adtSubscription: boolean,
```
Good suggestion. Fixed.

lucasdellabella commented on packages/api-sdk/src/medical/client/metriport.ts:
```diff
@@ -341,15 +341,17 @@ export class MetriportMedicalApi {
     facilityId: string,
-    additionalQueryParams: Record<string, string | number | boolean> = {}
+    additionalQueryParams: Record<string, string | number | boolean> = {},
+    adtSubscription = false
```
Would love to have an optional settings object, and keep adtSubscription as an optional field within. IMO the create patient interface should keep extra details like subscriptions and settings in an object to help the user of our SDK focus on the standard use case - creating a plain patient.

RamilGaripov replied:
Implemented

lucasdellabella reviewed (changes requested):
Just a few more small changes, looks solid!

leite08 commented on packages/api/src/sequelize/migrations/2025-03-13_00_create-patient-settings.ts:
```diff
@@ -0,0 +1,45 @@
+          field: adtSubscriptionColumn,
+          allowNull: false,
+          defaultValue: false,
+        },
```
We can still enforce schema type w/ the type in memory. Regardless of whether we have the JSONB or columns, we need to address existing records vs. required data.

I don't like JSONB when it comes to usability (writing queries), it slows me down. But it's super useful to update the schema w/o requiring DB migrations, which makes development/releases faster (and less risky b/c of https://github.com/metriport/metriport-internal/issues/920).

I wouldn't be too picky about this, but in this case we know that we'll add more settings, especially more HL7v2 notifications.

If we really want to keep it as a separate column, I'd suggest prefixing w/ subscription instead of using it as a suffix - easier to mentally group them together. Something like:
- `subscriptionAdt`
- `subscriptionRoc`
- `subscriptionOml`

lucasdellabella replied:
No after you pointed out that its hard to rollback migrations in cloud env im pretty in support of jsonb column. I also would still actually prefer we use the `subscriptionX` format to be used for keys. There will be other patient level settings, potentially many subscriptions that we enable by cx / patient. not all of them active subscriptions, and you're right, there's a good chance we might have other hl7 message types for patients.

leite08 commented on packages/api/src/sequelize/migrations/2025-03-13_00_create-patient-settings.ts:
```diff
@@ -3,7 +3,9 @@ import type { Migration } from "..";
-const adtSubscriptionColumn = "adt_subscription";
+const adtSubscriptionColumn = "subscribe_to";
+const patientSettingsTableConstraintName = "patient_settings_cxId_patientId_constraint";
+const patientSettingsTableIdFields = ["patient_id", "cx_id"];
```
Any particular reason to have this order?

IIRC, the order we specify these will reflect the order of the columns in the index, so we should be careful based on the usage we expect here.
- cx_id, patient_id:
  - searching for cx_id only might still allow the index to be used
  - does not allow to search by patient_id only
- patient_id, cx_id:
   - searching for cx_id only doesn't allow the index to be used
   - allows to search by patient_id only

RamilGaripov replied:
Oh, I wasn't aware of this at all! Thx for exposing this to me. So, in this case, `cx_id, patient_id` is definitely the preferred option

RamilGaripov commented on packages/api/src/command/medical/patient/get-adt-subscribers.ts:
```diff
@@ -0,0 +1,27 @@
+  const { log } = out(`Get ADT Subscribers`);
+  log(`States: ${states}`);
+
+  const patientIds = await PatientSettingsModel.findAll({
+    where: {
+      subscribeTo: { adt: true },
+    },
+    attributes: ["patientId"],
+  });
+
+  const patients = await PatientModel.findAll({
+    where: {
+      id: patientIds.map(p => p.patientId),
+    },
+  });
```
shit, this file shouldnt be in this PR either! dang it... This is from part 2 😒

leite08 commented on packages/api/src/command/medical/patient/create-patient-settings.ts:
```diff
@@ -0,0 +1,172 @@
+  patientIds,
+  subscribeTo = { adt: false },
+}: PatientSettingsUpsertProps): Promise<PatientSettingsUpsertResults> {
+  const { log } = out(`createOrUpdatePatientSettings - cx ${cxId}`);
```
update log prefix w/ fn name please

RamilGaripov commented on packages/api/src/command/medical/patient/create-patient-settings.ts:
```diff
@@ -36,56 +51,66 @@ export async function createPatientSettings({
+  cxId,
+  subscribeTo = { adt: false },
+}: PatientSettingsUpsertForCxProps): Promise<PatientSettingsUpsertResults> {
+  const { log } = out(`bulkUpsertPatientSettingsForCx - cx ${cxId}`);
```
Woops. I've been changing the func names back and forth, and forgot to update this line

lucasdellabella replied:
all good, always happens to me in reviews

leite08 commented on packages/api/src/command/patient-settings/deleteSettings.ts:
Please move this to `medical/patient/settings/`

leite08 replied:
We still have deleteSettings on its own folder, outside of medical/patient.

And we name files w/ kebab-case

leite08 commented on packages/api/src/command/medical/patient/create-patient-settings.ts:
Please move this to `medical/patient/settings/`

leite08 replied:
not moved yet?

leite08 commented on packages/api/src/command/medical/patient/create-patient-settings.ts:
```diff
@@ -0,0 +1,172 @@
+    id: existingSettingsMap.get(patientId)?.id ?? uuidv7(),
+    cxId,
+    patientId,
+    subscribeTo,
```
Since this is an object, we should prob build this code considering we might pass a single property and want to keep the existing values for other properties, right? Like a PATCH.

It's _very_ likely that we won't remember this function when we add the next property on `Subscriptions`.

RamilGaripov replied:
Ah, very good point!

leite08 commented on packages/api/src/command/medical/patient/create-patient-settings.ts:
```diff
@@ -0,0 +1,172 @@
+    subscribeTo,
+  }));
+
+  await PatientSettingsModel.bulkCreate(upserts, {
+    updateOnDuplicate: ["subscribeTo"],
+  });
```
Not sure what the default value is, but we should make sure we're not returning unnecessary data from the DB - we can use `returning: false` for this. And I think it makes it clearer to have the property there, even if the default is false.

See `createOrUpdateCoverageEnhancements()` for ref.

leite08 commented on packages/api/src/routes/medical/internal-patient.ts:
```diff
@@ -910,6 +914,63 @@ router.post(
+  asyncHandler(async (req: Request, res: Response) => {
+    const cxId = getUUIDFrom("query", req, "cxId").orFail();
+    const patientIds = getFromQueryAsArrayOrFail("patientIds", req);
+    const adtSubscription = getFromQueryAsBoolean("adtSubscription", req) ?? false;
```
Another instance of us defining the default value in multiple places. Let's prefer to do that on the command, please.

leite08 reviewed (commented):
Still missing updating Docs

leite08 commented on packages/api/src/routes/medical/internal-patient.ts:
```diff
@@ -941,6 +944,33 @@ router.post(
+ * @returns 200 with the results of the operation.
+ */
+router.post(
+  "/settings-bulk",
```
Also, probably a good idea to have a dedicated route file - it's already too large

RamilGaripov replied:
Moved

leite08 commented on packages/api-sdk/src/medical/models/patient-settings.ts:
```diff
@@ -0,0 +1,17 @@
+
+export const patientSettingsSchema = z.object({
+  settings: patientSettingsDataSchema.optional(),
+});
```
nit: I suggest we rename `patientSettingsDataSchema` to `patientSettingsSchema` and have the...

```ts
z.object({
  settings: patientSettingsDataSchema.optional(),
})
```

...as part of the `patientCreateSchema`, like we have the `externalId`.

This allows for the schema be named according with what it represents - there's no "data" patient settings.

leite08 commented on packages/api/src/command/medical/patient/settings/create-patient-settings.ts:
```diff
@@ -0,0 +1,182 @@
+    id: existingSettingsMap.get(patientId)?.id ?? uuidv7(),
+    cxId,
+    patientId,
+    subscriptions: {
+      ...existingSettingsMap.get(patientId)?.subscriptions,
+      ...settings.subscriptions,
+    },
```
Don't we have `settings` as the property that contains `subscriptions` inside each PatientSettings?

RamilGaripov replied:
Will address this once this is discussed/resolved - [link](https://github.com/metriport/metriport/pull/3463#discussion_r2001362550)

leite08 commented on packages/api/src/models/patient-settings.ts:
```diff
@@ -0,0 +1,35 @@
+  implements PatientSettings
+{
+  static NAME = "patient_settings";
+  declare id: string;
+  declare cxId: string;
+  declare patientId: string;
+  declare subscriptions: Subscriptions | undefined;
+
+  static setup: ModelSetup = (sequelize: Sequelize) => {
+    PatientSettingsModel.init(
+      {
+        ...BaseModel.attributes(),
+        cxId: {
+          type: DataTypes.UUID,
+        },
+        patientId: {
+          type: DataTypes.UUID,
+        },
+        subscriptions: {
+          type: DataTypes.JSONB,
+        },
+      },
```
This is not matching the domain/interface, which expects a `settings` prop in the body of the Model.

RamilGaripov replied:
Hmm.. I didn't want the settings to be the column on this table, but now I'm not sure.. I left it like this on purpose - if more settings attributes are added in the future, maybe we'd want to put those on new other columns in the table, but keep the subscriptions column just for subscriptions. Wdyt?

lucasdellabella commented on packages/api/src/routes/internal/internal.ts:
```diff
@@ -46,13 +46,15 @@ import hieRoutes from "../medical/internal-hie";
 
 router.use("/docs", docsRoutes);
 router.use("/patient", patientRoutes);
+router.use("/patient-settings", patientSettingsRoutes);
```
thought this was gonna be at route /patient/settings

RamilGaripov commented on packages/api-sdk/src/index.ts:
I actually reverted the reordering of imports cuz it looked so messy. Just adding this there:
```
export { patientSettingsDataSchema } from "./medical/models/patient-settings";
```

leite08 commented on packages/api/src/command/medical/patient/settings/create-patient-settings.ts:
```diff
@@ -0,0 +1,182 @@
+
+  await PatientSettingsModel.bulkCreate(upserts, {
+    returning: false,
+    updateOnDuplicate: ["subscriptions"],
```
ditto: `settings.subscriptions`?

leite08 commented on packages/api/src/sequelize/migrations/2025-03-18_00_create-patient-settings.ts:
```diff
@@ -0,0 +1,59 @@
+import * as shared from "../migrations-shared";
+
+const patientSettingsTableName = "patient_settings";
+const adtSubscriptionColumn = "subscriptions";
```
supernit: rename the const to remove `adt`

leite08 commented on packages/api/src/command/medical/patient/settings/create-patient-settings.ts:
```diff
@@ -0,0 +1,182 @@
+ */
+export async function upsertPatientSettingsForPatientList({
+  cxId,
+  facilityId,
```
In this case, since we're passing the list of patient IDs, maybe we don't have the facility ID?

RamilGaripov replied:
It's an optional param. I think it's fine if we want that to be an extra guard rail when updating patients for a certain facility, no?

RamilGaripov commented on packages/api/src/command/medical/patient/settings/create-patient-settings.ts:
```diff
@@ -0,0 +1,184 @@
+  return { patientsFoundAndUpdated };
+}
+
+async function verifyPatients({
+  patientIds,
+  facilityId,
+  cxId,
+}: {
+  patientIds: string[];
+  facilityId?: string;
+  cxId: string;
+}): Promise<{
+  validPatientIds: string[];
+  invalidPatientIds: string[];
+}> {
+  if (patientIds.length < 1) {
+    return {
+      validPatientIds: [],
+      invalidPatientIds: [],
+    };
+  }
+
+  const patients = await PatientModel.findAll({
+    where: { id: patientIds, cxId, ...(facilityId && { facilityIds: [facilityId] }) },
+    attributes: ["id"],
+  });
+  const foundPatientIds = new Set(patients.map(p => p.id));
+  const invalidPatientIds = patientIds.filter(id => !foundPatientIds.has(id));
+  return {
+    validPatientIds: Array.from(foundPatientIds),
+    invalidPatientIds,
+  };
+}
+
```
Good catch!

leite08 commented on packages/api/src/command/medical/patient/settings/create-patient-settings.ts:
```diff
@@ -0,0 +1,184 @@
+ * @param settings Patient settings object, which includes subscriptions
+ * @returns The number of patients updated and the list of patients not found
+ */
+export async function upsertPatientSettingsForPatientList({
+  cxId,
+  facilityId,
+  patientIds,
+  settings,
+}: PatientSettingsUpsertProps): Promise<PatientSettingsUpsertResults> {
+  const { log } = out(`upsertPatientSettingsForPatientList - cx ${cxId}`);
+
+  const { validPatientIds, invalidPatientIds: patientsNotFound } = await verifyPatients({
+    cxId,
+    facilityId,
+    patientIds,
+  });
+
+  if (validPatientIds.length === 0) {
+    throw new BadRequestError(`No valid patients found`);
+  }
+
+  const patientsFoundAndUpdated = await upsertPatientSettings({
+    cxId,
+    patientIds: validPatientIds,
+    settings,
+  });
+
+  log(`Updated settings for ${patientsFoundAndUpdated} patients`);
+  return { patientsFoundAndUpdated, patientsNotFound };
+}
```
This is not a requirement in this case, as we can just re-execute the `upsertPatientSettingsForPatientList` and it will attempt to leave the system on the same state - it's an idempotent operation.

lucasdellabella reviewed (approved):
🚢

leite08 commented on packages/api/src/routes/internal/medical/patient-settings.ts:
```diff
@@ -0,0 +1,78 @@
+ * @param req.query.patientIds List of patient IDs to update.
+ * @returns 200 with the results of the operation.
+ */
+router.post(
+  "/",
+  requestLogger,
+  asyncHandler(async (req: Request, res: Response) => {
+    const cxId = getUUIDFrom("query", req, "cxId").orFail();
+    const facilityId = getUUIDFrom("query", req, "facilityId").optional();
+    const patientIds = getFromQueryAsArrayOrFail("patientIds", req);
+    const settings = patientSettingsSchema.parse(req.body) ?? defaultSettings;
+
+    const result = await upsertPatientSettingsForPatientList({
+      cxId,
+      facilityId,
+      patientIds,
+      settings,
+    });
+
+    return res.status(status.OK).json(result);
+  })
+);
```
Our default Express error handler (see `packages/api/src/routes/helpers/default-error-handler.ts`) already processes Zod schema validation errors and return a 4xx to the client. This is our pattern for routes, no need for a special handler.

leite08 commented on packages/api/src/routes/internal/medical/patient-settings.ts:
```diff
@@ -0,0 +1,78 @@
+ * @param req.query.facilityId The facility ID. Optional.
+ * @returns 200 with the results of the operation.
+ */
+router.post(
+  "/bulk",
+  requestLogger,
+  asyncHandler(async (req: Request, res: Response) => {
+    const cxId = getUUIDFrom("query", req, "cxId").orFail();
+    const facilityId = getUUIDFrom("query", req, "facilityId").optional();
+    const settings = patientSettingsSchema.parse(req.body) ?? defaultSettings;
+
+    const result = await upsertPatientSettingsForCx({
+      cxId,
+      facilityId,
+      settings,
+    });
+
+    return res.status(status.OK).json(result);
+  })
+);
```
Same as https://github.com/metriport/metriport/pull/3463#discussion_r2002221429

leite08 commented on packages/api/src/command/medical/patient/settings/create-patient-settings.ts:
```diff
@@ -26,6 +27,14 @@ type PatientSettingsUpsertProps = PatientSettingsUpsertForCxProps & {
+type BatchProcessingResult = {
+  processedCount: number;
+  failedCount: number;
+  error?: Error;
```
I suggest we treat those as `unknown`. Almost sure we're getting Error, but in JS/TS it's valid to send other types as errors.

leite08 commented on packages/api/src/command/medical/patient/settings/create-patient-settings.ts:
```diff
@@ -89,8 +98,33 @@ export async function upsertPatientSettingsForPatientList({
+  cxId,
+  settings,
+}: {
+  patients: string[];
```
please rename `patients` to `patientIds`, it's clearer

leite08 commented on packages/api/src/command/medical/patient/settings/create-patient-settings.ts:
```diff
@@ -89,8 +98,33 @@ export async function upsertPatientSettingsForPatientList({
 }
 
+/**
+ * Processes a batch of patient settings updates
```
nit: updates > upserts

leite08 commented on packages/api/src/command/medical/patient/settings/create-patient-settings.ts:
```diff
@@ -104,21 +138,51 @@ export async function upsertPatientSettingsForCx({
+  do {
+    const { currPageItems, nextPageItemId } = await getPaginationItems(
+      { count: BATCH_SIZE, ...(fromItem ? { fromItem } : {}) },
+      pagination => getPatients({ cxId, facilityId, pagination }),
```
Before we were only loading the patient IDs, now we're loading the whole patients.

Couldn't we just keep the previous version that gets all the patient IDs and then batches only the upserts?

IMO the issue was not with loading a lot of IDs from the DB, but with the upsert of those into the DB.

For reference, I populated an array w/ 1MM UUIDs in 5 seconds on local.

RamilGaripov commented on packages/api/src/command/medical/patient/settings/create-patient-settings.ts:
```diff
@@ -0,0 +1,279 @@
+  failedIds?: string[];
+};
+
+const BATCH_SIZE = 1;
```
Woops! Forgot to change after testing locally with a small number of patients

leite08 commented on packages/api/src/command/medical/patient/settings/create-patient-settings.ts:
```diff
@@ -30,10 +32,15 @@ type PatientSettingsUpsertProps = PatientSettingsUpsertForCxProps & {
   processedCount: number;
   failedCount: number;
-  error?: Error;
+  failedIds?: string[];
```
Count is not necessary as we have the list of failed IDs, right?

RamilGaripov replied:
It doesn't hurt tho, right? Since we'll be working with large numbers of pts per CX, wouldnt it be nice to see the number at a glance rather than going thru the extra step to get the count?

leite08 replied:
By having both we introduce the possibility of them not matching. If we can have single source of truth, I think it's better, don't you think?

RamilGaripov replied:
Hmm. I see what you mean in a bigger picture sorta way, but in this case, I'd prefer to keep it. This is used for internal operations, and I think these numbers will match here + it's quick feedback on the results of an operation that would require attention.

leite08 commented on packages/api/src/command/medical/patient/settings/create-patient-settings.ts:
```diff
@@ -99,32 +106,48 @@ export async function upsertPatientSettingsForPatientList({
+      });
+      return { processedCount, failedCount: 0 };
+    } catch (error) {
+      const isLastAttempt = attempt === MAX_BATCH_RETRIES;
```
nit: try to use "defensive programming", i.e., `attempt >= MAX_BATCH_RETRIES`

leite08 replied:
@coderabbitai flag cases where we can use defensive programming like this, especially around counts and comparison with `undefined` (== vs === so it also checks for `null`).

leite08 replied:
> wait so we prefer == over ===? I thought we were striving for ===?

We always should use `===` to compare stuff, with the exception of comparing with `undefined`. The reason is that for our purposes, `null` and `undefined` are the same, but if we compare w/ `===` they'll be treated differently.

By using `==` to check for `undefined`, we also include `null`.

I don't like the idea to have this exception as it requires additional explanation (as we're doing now), but so far have considered this as a price to pay to have code a bit more robust (sometimes we get stuff in TS/JS that's not what we expect).

Happy to discuss more on Slack!

leite08 commented on packages/api/src/command/medical/patient/settings/create-patient-settings.ts:
```diff
@@ -99,32 +106,48 @@ export async function upsertPatientSettingsForPatientList({
+    } catch (error) {
+      const isLastAttempt = attempt === MAX_BATCH_RETRIES;
+
+      if (!isLastAttempt) {
+        log(
+          `Batch operation failed, attempt ${attempt}/${MAX_BATCH_RETRIES}: ${errorToString(error)}`
+        );
+      }
```
NABD, but since we're logging this, might as well add a log for the last attempt, saying it error'd and we're not trying again.

leite08 replied:
NABD, but since we're logging this, might as well add a log for the last attempt, saying it error'd and we're not trying again.

leite08 commented on packages/api/src/command/medical/patient/settings/create-patient-settings.ts:
```diff
@@ -138,50 +161,52 @@ export async function upsertPatientSettingsForCx({
-    const patientIds = currPageItems.map(p => p.id);
-    const { processedCount, failedCount, error } = await processBatch({
-      patients: patientIds,
+  for (let i = 0; i < patientIds.length; i += BATCH_SIZE) {
+    const batchIds = patientIds.slice(i, i + BATCH_SIZE);
```
No need to update this now, but it's easier (and cleaner) to use Lodash's [.chunk()](https://lodash.com/docs/4.17.15#chunk) - recommending this because we use Lodash throughout the codebase, including a bunch of usages of `.chunk()`

leite08 commented on packages/api/src/command/medical/patient/settings/create-patient-settings.ts:
```diff
@@ -138,50 +161,52 @@ export async function upsertPatientSettingsForCx({
+
+  if (failedTotal > 0) {
+    const msg = `Failed to upsert settings for patients`;
+    log(`${msg} - failed IDs: ${failedIds}`);
```
Is this outputting the list of IDs, or just `[object]`?

leite08 replied:
@coderabbitai make sure to flag cases where we're sending objects to string templates without converting them to string first

leite08 commented on packages/api/src/command/medical/patient/settings/create-patient-settings.ts:
```diff
@@ -138,50 +161,52 @@ export async function upsertPatientSettingsForCx({
+    log(`${msg} - failed IDs: ${failedIds}`);
+    capture.error(msg, {
+      extra: {
+        failedIds,
+        cxId,
+        facilityId,
```
@coderabbitai the order or properties inside `capture.error` and `capture.message`'s `extra` parameter matters: this is passed on to Sentry, and if we send too much data, Sentry will trim out the data beyond it's payload limit. We need to be careful to start with the most important information, in this case the `cxId`. The `facilityId` really can be the last one (not in all cases, though).

@

leite08 commented on packages/api/src/routes/medical/internal-patient-settings.ts:
```diff
@@ -0,0 +1,78 @@
+
+    return res.status(status.OK).json(result);
+  })
+);
```
Done

leite08 commented on packages/api-sdk/src/medical/models/patient.ts:
We're missing the alpha version published.

@coderabbitai when there are changes in the "SDK" packages (`api-sdk`, `commonwell-sdk`, and `carequality-sdk`), as well as `packages/shared`, we need to deploy alpha versions of these on NPM.

leite08 commented on packages/api/src/models/patient-settings.ts:
```diff
@@ -20,7 +20,14 @@ export class PatientSettingsModel
         patientId: {
-          type: DataTypes.UUID,
+          type: DataTypes.STRING,
+          references: {
+            model: "patient",
+            key: "id",
+          },
+          onDelete: "CASCADE",
+          onUpdate: "CASCADE",
+          allowNull: false,
```
Do we need to have these on the Model file? AFAIK we only keep here the bare minimum info, with the more DB low-level stuff on the migration file (like `allowNull`).

RamilGaripov replied:
I'll take out these fields:
- onDelete: "CASCADE",
- onUpdate: "CASCADE",
- allowNull

But we still need the references to allow us to use `include` operations on DB queries in the app
